### PR TITLE
DDS-1020 Add Tab Component

### PR DESCRIPTION
**Pull request by yodas7**
This PR adds Navigation Tab (`daikin-tab` and `daikin-tab-group`) component.

## Related Issue(s) and Links

- JIRA Ticket (parent): [DDS-1020](https://daikinsvoil.atlassian.net/browse/DDS-1020)

## API Design

- [JIRA Wiki for Tab Component](https://daikinsvoil.atlassian.net/wiki/spaces/GDS/pages/3018686539/Tab+Component+API)
- [JIRA Wiki for Tab Group Component](https://daikinsvoil.atlassian.net/wiki/spaces/GDS/pages/3042738180/Tab+Group+Component+API)

## How Has This Been Tested?

- VRT (Visual Regression Test) for \[Default; Condensed\] size variants and \[Normal; Hover; Pressed; Disabled\] states matrix.
- VRT: Scroll into active tab on initial render
- Interaction test: Changes active tab when another enabled tab clicked
- Interaction test: Does not change active tab when disabled tab clicked
- Interaction test: The panel shows the content associated with the active tab
- Interaction test: (Completely navigate-able only with keyboard

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have added appropriate VRTs.
- [x] I have added appropriate interaction tests.
- [x] I have confirmed that my changes do not generate new warnings and errors.

[DDS-1020]: https://daikinsvoil.atlassian.net/browse/DDS-1020?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by yodas7** [comment]
CI is failing due to text rendering.
Since we are using an M1 machine, a GitHub Actions M1 runner may solve the problem, but this always seems to be paid for.
I'll try setting `text-rendering`.

File: src/components/tab/daikin-tab-group.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  private _managingTabs: [DaikinTab, () => void][] = [];
+
+  private _handleBeforeChange(newValue: string): boolean {
+    if (this.value === newValue) {
+      return false;
+    }
+
+    if (
+      !this.dispatchEvent(
+        new CustomEvent("beforechange", {
```
Comment (KoharaKazuya) [inline src/components/tab/daikin-tab-group.ts:34]: @yodas7 
Is daikin-` prefix needed? Please track this discussion. https://github.com/dsv-rp/DDS/pull/4#discussion_r1653837344

File: src/components/tab/stories/daikin-tab-group-wc.stories.ts
Comment (KoharaKazuya) [inline src/components/tab/stories/daikin-tab-group-wc.stories.ts:1]: @yodas7 
[IMO] It would be great if we have stories for 0, 1 and many tabs (to check scroll function).

File: __test__/__image_snapshots__/button-test-js-button-primary-hover-1-snap.png
Comment (KoharaKazuya) [inline __test__/__image_snapshots__/button-test-js-button-primary-hover-1-snap.png:1]: @yodas7 
Is this an expected change?

**Review by KoharaKazuya — CHANGES_REQUESTED** [review]

File: jest-puppeteer.config.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// Ddd this code to improve chrome browser rendering performance when test
```
[the remark this answers is not in the document]
Comment (yodas7) [reply jest-puppeteer.config.js]: Fixed by merging develop branch.

File: __test__/setupTests.js [hunk 3]
```diff
@@ -1,8 +1,17 @@
-const { configureToMatchImageSnapshot } = require('jest-image-snapshot');
+const { configureToMatchImageSnapshot } = require("jest-image-snapshot");
 
 const toMatchImageSnapshot = configureToMatchImageSnapshot({
-    failureThreshold: 0.01,
-    failureThresholdType: 'percent'
+  allowSizeMismatch: true,
+  failureThreshold: 0.1, // 10%, very large value; this will be changed after we run VRTs in a container
```
[the remark this answers is not in the document]
Comment (yodas7) [reply __test__/setupTests.js]: Added comment

File: __test__/__image_snapshots__/button-test-js-button-primary-hover-1-snap.png
Comment (yodas7) [reply __test__/__image_snapshots__/button-test-js-button-primary-hover-1-snap.png:1]: Intended. Updated all snapshots by running `npm run test:update`.
FYI we'll do pixel-perfect VRT in #8 so that we don't have to concern about snapshots even more.

File: src/components/tab/stories/daikin-tab-group-wc.stories.ts
Comment (yodas7) [reply src/components/tab/stories/daikin-tab-group-wc.stories.ts:1]: Added except for 0 tabs.  
Tab group component requires at least one active tab exists in children (wrote comment on component class JSDoc).

Comment (KoharaKazuya) [reply src/components/tab/stories/daikin-tab-group-wc.stories.ts:1]: Thanks

File: __test__/__image_snapshots__/button-test-js-button-primary-hover-1-snap.png
Comment (KoharaKazuya) [reply __test__/__image_snapshots__/button-test-js-button-primary-hover-1-snap.png:1]: I got it. No problem.

File: __test__/setupTests.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (KoharaKazuya) [reply __test__/setupTests.js]: Thanks!

**Comment by yodas7** [comment]
@KoharaKazuya @hiro-daikin 
Could you review this PR?

File: src/components/tab-group/daikin-tab-group.ts [hunk 4]
```diff
@@ -0,0 +1,223 @@
+import { LitElement, html, css, unsafeCSS } from "lit";
+import { customElement, property } from "lit/decorators.js";
+
+import tailwindStyles from "../../tailwind.css";
+import type { DaikinTab } from "../tab/daikin-tab";
+
+/**
+ * Tab Group that automatically manages child tabs.
+ *
+ * NOTE: If there is no tab available in the child elements, i.e., zero tabs, or if all tabs are disabled, it is ILLEGAL.
```
Comment (KoharaKazuya) [inline src/components/tab-group/daikin-tab-group.ts:10]: @yodas7 
Does `ILLEGAL` mean undefined behavior? In my opinion, for safety, we should define some behavior. Any behavior is fine. What do you think?

Comment (yodas7) [reply src/components/tab-group/daikin-tab-group.ts:10]: @KoharaKazuya 
Undefined behavior is correct.
I'd like to define "undefined behavior" to include unintended display, generation of unintended values, and throwing of exceptions, but not freezing.
In practice, it will probably remain unchanged as it was effective in the past, but the library does not guarantee any behavior and the behavior may change due to version upgrades.

**Review by KoharaKazuya — APPROVED** [review]

**Comment by yodas7** [comment]
@rinjInTokyo invented a drastic improvement that eliminates the need to attach listeners by bubbling events. I will not change it in this PR as it has already been reviewed, but will refactor it in the future.

File: src/components/tab/daikin-tab.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+  override focus(options?: FocusOptions | undefined): void {
+    this.shadowRoot?.querySelector("button")?.focus(options);
+  }
+
+  render() {
+    const cn = [
+      STYLE_COMMON,
+      STYLE_MAP_SIZES[this.size ?? "default"] ?? STYLE_MAP_SIZES.default,
+    ].join(" ");
```
Comment (hiro-daikin) [inline src/components/tab/daikin-tab.ts:113]: Any reason why we're not using cva? 
We can also use it to automatically get types

File: src/components/tab-group/daikin-tab-group.ts (hunk 4 above)
Comment (hiro-daikin) [inline src/components/tab-group/daikin-tab-group.ts:10]: Should we throw an error for these edge cases?

**Review by hiro-daikin — CHANGES_REQUESTED** [review]

File: __test__/Tab.test.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    // wait for element to be visible
+    const tab = await page.waitForSelector("daikin-tab", { visible: true });
+
+    // check if the internal button element is not disabled
+    const isDisabled = await page.evaluate(
+      (tab) => tab.shadowRoot.querySelector("button").disabled,
+      tab
+    );
+
+    expect(isDisabled).toBe(false);
```
[the remark this answers is not in the document]
Comment (yodas7) [reply __test__/Tab.test.js:109]: We need to implement a tab panel that is not yet in this PR.
I'll try to implement a tab panel on this PR later.

File: src/utils/decoratorListener.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import type { LifecycleMixin, LifecycleMixinStatic } from "./mixinLifecycle";
+
+export function on<K extends keyof HTMLElementEventMap>(
```
[the remark this answers is not in the document]
Comment (yodas7) [reply src/utils/decoratorListener.ts:3]: I'll add comments later.
I also feel that it would be better to create a base class rather than mixins, so I'll change that as well.

File: src/components/tab-group/daikin-tab-group.ts (hunk 4 above)
Comment (yodas7) [reply src/components/tab-group/daikin-tab-group.ts:10]: I'm not sure how best to inform users (developers) when a component is used in a way that it is not intended to be used.
For example, frameworks such as React only show errors during development and are silent in production builds.
Should we incorporate such a mechanism?

**Comment by yodas7** [comment]
I'd like to wait for #8 to be merged as both this PR and #8 may change component fundamentals.

File: src/components/tab/daikin-tab.ts (hunk 5 above)
Comment (yodas7) [reply src/components/tab/daikin-tab.ts]: Migrated to cva.

File: src/utils/mixinLifecycle.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+export interface LifecycleMixin<This> {
+  registerLifecycleCallback(callback: Registration<This>): void;
+}
+
+export interface LifecycleMixinStatic<This> {
+  _staticLifecycleRegistrationSet: Set<Registration<This>>;
+  registerStaticLifecycleCallback(callback: Registration<This>): void;
+}
+
+export function mixinLifecycle<T extends Constructor<LitElement>>(
```
[the remark this answers is not in the document]
Comment (yodas7) [reply src/utils/mixinLifecycle.ts]: These mixins and decorators proved unnecessary and were removed to reduce complexity.

File: src/utils/decoratorListener.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (yodas7) [reply src/utils/decoratorListener.ts]: These mixins and decorators proved unnecessary and were removed to reduce complexity.

File: __test__/Tab.test.js [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    // wait for element to be visible
+    const tab = await page.waitForSelector("daikin-tab", { visible: true });
+
+    // check if the internal button element is disabled
+    const isDisabled = await page.evaluate(
+      (tab) => tab.shadowRoot.querySelector("button").disabled,
+      tab
+    );
+
+    expect(isDisabled).toBe(true);
```
[the remark this answers is not in the document]
Comment (yodas7) [reply __test__/Tab.test.js]: Updated.

File: __test__/Tab.test.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (yodas7) [reply __test__/Tab.test.js]: Implemented.

File: src/components/tab-group/daikin-tab-group.ts (hunk 4 above)
Comment (yodas7) [reply src/components/tab-group/daikin-tab-group.ts]: lit-html seems to use `development` field of `exports` in [package.json](https://github.com/lit/lit/blob/main/packages/lit-html/package.json#L15-L28).
ref: https://nodejs.org/api/packages.html#community-conditions-definitions

Ultimately, I think it would be best to provide both development and production code, as lit-html does, but I'm not sure at what point we should do that.
As for building, it is easy to achieve as it is just a `vite build` twice with different modes.

Comment (hiro-daikin) [reply src/components/tab-group/daikin-tab-group.ts]: @yodas7 Since one goal of a design system is to have uniform design/UI across projects, I think it's okay to throw the error in both development and production. 

If we aren't strict in the beginning, it would be extremely hard to change later. 
If there is a use case which isn't covered, the user can file an issue/ticket.

Comment (yodas7) [reply src/components/tab-group/daikin-tab-group.ts]: @hiro-daikin 
Shortly after I wrote my comment, I changed it to include development builds.
Sorry for not notifying you.
I would appreciate it if you could check to see if this change looks good.

File: package.json [hunk 10]
```diff
@@ -9,23 +9,42 @@
...
+  "main": "./dist/cjs/index.cjs",
   "module": "./dist/es/index.js",
   "types": "./dist/es/index.d.ts",
   "exports": {
     ".": {
+      "default": "./dist/es/index.js",
       "import": "./dist/es/index.js",
-      "require": "./dist/cjs/index.cjs"
+      "require": "./dist/cjs/index.cjs",
+      "development": {
```
[the remark this answers is not in the document]
Comment (yodas7) [reply package.json:20]: The official information is only on the [Webpack website](https://webpack.js.org/guides/package-exports/#providing-devtools-or-production-optimizations).
[Vite also supports `development` condition](https://ja.vitejs.dev/config/shared-options#resolve-conditions).

- About nested conditions: https://nodejs.org/api/packages.html#nested-conditions
- Known community conditions : https://nodejs.org/api/packages.html#community-conditions-definitions

[the remark this answers is not in the document]
Comment (yodas7) [reply package.json:20]: A practical examples can be found here

- https://github.com/lit/lit/blob/main/packages/lit-html/package.json
- https://github.com/vuejs/core/blob/main/packages/runtime-dom/package.json

[the remark this answers is not in the document]
Comment (KoharaKazuya) [reply package.json:20]: Thank you! I see.

**Review by KoharaKazuya — APPROVED** [review]

File: .storybook/react/main.ts [hunk 11]
```diff
@@ -1,8 +1,15 @@
...
 import { env } from "node:process";
-import { getAllStorybookFiles, STORYBOOK_ADDONS } from "../main-common";
+import {
+  getAllStorybookFiles,
+  STORYBOOK_ADDONS,
+  viteFinal,
+} from "../main-common";
 
-env.STORYBOOK_FW = "react";
+if (!env.STORYBOOK_MAIN_LOADED) {
```
Comment (hiro-daikin) [inline .storybook/react/main.ts:9]: Can you add comments for this block?

File: build/vite/storybook-framework-loader.ts [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+  ...fwMetadata,
+  parameters: {
+    docs: {
+      description: {
+        component: ${JSON.stringify(componentDescription)},
+      }
+    }
+  },
+};
+`;
```
Comment (hiro-daikin) [inline build/vite/storybook-framework-loader.ts:68]: There seems to be some formatting issues here. 
Does prettier not catch template strings?

File: package.json [hunk 13]
```diff
@@ -9,23 +9,42 @@
...
+        "import": "./dist/es-dev/*.js",
+        "require": "./dist/cjs-dev/*.cjs"
+      }
     }
   },
   "sideEffects": true,
   "scripts": {
     "prepare": "husky",
-    "build": "tsc -b && vite build",
+    "build": "tsc -b && vite build && vite build --mode development",
```
Comment (hiro-daikin) [inline package.json:47]: I'm assuming this builds both for development and production?
Would it be better to separate this out to a `build:dev` script?

Main concern is it would take twice as long to build for production (such as in CICD pipeline)

File: scripts/storybook-ci.ts [hunk 14]
```diff
@@ -16,11 +18,25 @@ const command = {
   },
 }[STORYBOOK_FW][STORYBOOK_ENV];
 
-console.info(`> ${command}\n`);
+if (STORYBOOK_ENV === "development") {
+  console.error("Removing node_modules/.cache/storybook...");
```
Comment (hiro-daikin) [inline scripts/storybook-ci.ts:22]: Does this prevent you from proceeding?
If not, would `console.info` or `console.warn` be more appropriate?

File: src/components/tab/stories/daikin-tab.stories.ts [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+  play: definePlay(async ({ args, canvasElement, step }) => {
+    const root = canvasElement.getElementsByTagName("daikin-tab")[0];
+    await expect(root).toBeInTheDocument();
+
+    const innerButton = getByShadowRole(root, "tab", {
+      name: "Tab",
+    });
+    await expect(innerButton).toBeInTheDocument();
+
+    // should react if inner button clicked
```
Comment (hiro-daikin) [inline src/components/tab/stories/daikin-tab.stories.ts:71]:
```suggestion
    // should not react if inner button clicked
```

File: src/components/tab/stories/daikin-tab.stories.ts [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    await step("Try to click inner button", async () => {
+      await userEvent.click(innerButton);
+      await expect(args.onClick).not.toHaveBeenCalled();
+    });
+
+    // should also react if outer button clicked
```
Comment (hiro-daikin) [inline src/components/tab/stories/daikin-tab.stories.ts:77]:
```suggestion
    // should also not react if outer button clicked
```

**Review by hiro-daikin — CHANGES_REQUESTED** [review]

File: src/components/tab/stories/daikin-tab.stories.ts (hunk 15 above)
Comment (yodas7) [reply src/components/tab/stories/daikin-tab.stories.ts:71]: THanks.

File: scripts/storybook-ci.ts (hunk 14 above)
Comment (yodas7) [reply scripts/storybook-ci.ts:22]: I wanted to output to stderr because Playwright throws all server side stdout logs away.
`console.error` is confusing, so I changed it to `stderr.write` and added a comment.

File: package.json (hunk 13 above)
Comment (yodas7) [reply package.json:47]: Yes, they are.

In certain cases, such as testing, you may only need one of the builds, but both must be present for the final publication.
I am concerned that by splitting up the commands, it makes it easier for `dist` to be incomplete.

File: build/vite/storybook-framework-loader.ts (hunk 12 above)
Comment (yodas7) [reply build/vite/storybook-framework-loader.ts:68]: Prettier will only format template strings under certain conditions (i.e., when it is safe to change the code).
I have addressed this by creating the object on the plugin side and making the code template simpler.

File: .storybook/react/main.ts (hunk 11 above)
Comment (yodas7) [reply .storybook/react/main.ts]: Moved this type of processing to main-common and wrote a comment there.

File: package.json (hunk 13 above)
Comment (hiro-daikin) [reply package.json:47]: I see your concern. 

I've seen some patterns where they split up into:
build:dev
build:prod
build (combination of above 2)

If it makes sense, we can also follow this sort of pattern.

I'll leave the final implementation up to you, however.

**Review by hiro-daikin — APPROVED** [review]

File: package.json (hunk 13 above)
Comment (yodas7) [reply package.json:47]: It only took 12 seconds to build, so I think I'll leave it as it is for now.
If the build itself starts taking longer in the future, I will consider separating it.
